elastic/elastic-agent
# Report FQDN on host.name as opt-in, enabled by feature flag

## What does this PR do?

This PR enhances Elastic Agent to understand feature flags configuration, supplied either via standalone policy or Fleet policy, and applies said configuration to itself as well as to its components. 



## Why is it important?

To implement the first feature flag, FQDN.  Enabling the FQDN feature flag will allow Agent and its components to report their host's fully-qualified domain name via the `host.name` field in events generated by the components.



## Checklist




## Author's Checklist

### Manual testing of Agent in Standalone mode

#### Without FQDN feature flag enabled

https://user-images.githubusercontent.com/51061/223605084-894dadf1-f7b7-42ca-93c4-0f6cdceb18dc.mov


#### With FQDN feature flag enabled

https://user-images.githubusercontent.com/51061/223605093-b4159b4e-d22b-42ba-808a-87c8bc95a165.mov


## How to test this PR locally

_The screen recordings in the section above were created by following the same steps as listed below._

1. Build Agent package.
2. Spin up a VM, setting hostname to a FQDN, e.g. `foo.bar.baz`.
3. Copy Agent package into VM.
4. Setup an ES + (optionally) Kibana deployment.
5. Setup two standalone Agent configurations, one with the FQDN feature flag not enabled and one with it enabled.  In both, point the output to the ES cluster in the deployment you created.
6. Start the Agent with the first configuration (FQDN feature flag not enabled).
7. After a few seconds, check ES (or Kibana Discover) for data.  Ensure that `host.name` and `host.hostname` are both set to the short hostname, e.g. `foo` in all documents.
8. Stop Agent.
9. Start the Agent with the second configuration (FQDN feature flag enabled).
10. After a few seconds, check ES (or Kibana Discover) for data.  Ensure that `host.name` is set to the FQDN, e.g. `foo.bar.baz` while `host.hostname` is set to the short hostname, e.g. `foo`, in all documents.
11. Stop Agent.



## Related issues

- Closes #2185 



## Use cases



## Screenshots



## Logs
